### DISCUSSION TITLE
Trying to select a non-visible tree node, what's Selection.SelectedIndex?

### DISCUSSION THREAD

**Discussion by BhaaLseN** [AvaloniaUI/Avalonia#17710] [Q&A]
I'm currently migrating a WPF application to Avalonia 11.x and got _almost_ everything done. The one thing that's missing is the tree sync after selecting a search result in a flat list.

I basically got a `TreeView` with node view models bound to it, and a `ListBox`/`TextBox` next to it that allow me to search for nodes. Selecting a result in the listbox should select the associated tree node, and expand the ancestors to make it fully visible (all based on the view model behind the `TreeViewItem`).
In WPF, I had to ask each levels `ItemsControl.ItemContainerGenerator` if its `Status` is loaded already, and if not attach to its `StatusChanged` event to keep selecting once the expansion is complete and UI items are present. Inside this, I simply used `ItemsControl.ItemContainerGenerator.ContainerFromIndex` to grab the associated container, set `IsExpanded` and continue until I hit the desired item.
In Avalonia, no such thing exists; it even refers me back from `ItemsControl.ItemContainerGenerator.ContainerFromIndex` to simply ask the parent `ItemsControl.ContainerFromIndex`. There's no expansion/visuals/control status, and I can't detect if and when a level is present, so I'm often stuck when the desired node is deeper in the tree (because the containers don't exist yet).

I've tried a few variations of #13475,  #11845 and #16485 but none of them _really_ get me where I want to be.
_(Don't really worry about my progress there, I might switch to a `TreeDataGrid` anyways, and do some other things to get this to work.)_

Then I stumbled over the remark on `TreeView.SelectedItem`:
https://github.com/AvaloniaUI/Avalonia/blob/18fcfcc16348401e74fd74abba11223e6f7b4fb0/src/Avalonia.Controls/TreeView.cs#L111-L112

And...well...what _is_ even `Selection.SelectedIndex`? I know my indices per level, and I can keep track of the global index (in a flattened list), but I couldn't figure out what this meant. There is a `Avalonia.Controls.Selection` namespace and a `AvaloniaEdit.Editing.Selection` class, but none of those really help me. What seemed more promising was `Avalonia.Controls.Selection.ISelectionModel` (which has a `SelectedIndex` property), but no connection to the `TreeView`.

To get to the point: Is that documentation comment outdated? Should it point us elsewhere, or is it obsolete and doesn't apply anymore?

With the information i currently have on this, I don't know how to proceed, and the only useful results I can find on Google is the source code comment and the documentation; so it feels like something that needs work (one way or another).

**Comment by msneijders** [comment]
>I basically got a TreeView with node view models bound to it

You could give your node viewmodel the properties `IsSelected` and `IsExpanded`, using a `INotifyPropertyChanged` principle. Then setting the correct selected and expanded state using your viewmodel (and expand all parent nodes of the selected one up until the root node). And attach your view model like this to bind your node to the `TreeViewItem` of the `TreeView`:

```xaml
<TreeView
    ItemsSource="{Binding Items}">
    <TreeView.ItemTemplate>
        <TreeDataTemplate ItemsSource="{Binding Children}">
            <TextBlock Text="{Binding Header}" />
        </TreeDataTemplate>
    </TreeView.ItemTemplate>
    <TreeView.ItemContainerTheme>
        <ControlTheme
            x:DataType="viewModels:Node"
            BasedOn="{StaticResource {x:Type TreeViewItem}}"
            TargetType="TreeViewItem">
            <Setter Property="IsSelected" Value="{Binding IsSelected}" />
            <Setter Property="IsExpanded" Value="{Binding IsExpanded}" />
        </ControlTheme>
    </TreeView.ItemContainerTheme>
</TreeView>
```

**Reply by BhaaLseN** [reply]
Yeah, I got this part figured out, don't worry about it. But thanks anyways.

This discussions is about potentially missing documentation, an incorrect comment, or some other thing that needs to be addressed (but I didn't want to create an issue for it straight away, I wanted to see if this is _actually_ something to look at or not).

**Reply by msneijders** [reply]
ok -- about that remark about `Selection.SelectedIndex`. I think you are right and it is misplaced there, as I suspect it refers to `Selection` of `SelectingItemsControl`, but TreeView is not a SelectingItemsControl.

**Reply by timunie** [reply]
PR to fix documentation are welcome 🤗

**Reply by Flithor** [reply]
Something wrong about this answer.
If the TreeView is recreated (such as in flyout), selected item IsSelected will be reset to false when the TreeViewItem created.

<del>I don't know there is a bug or just something binding issue.</del>

Progress:
It's caused by line 415:
https://github.com/AvaloniaUI/Avalonia/blob/934c0f834946035d9b9bfd873739cf5c880de868/src/Avalonia.Controls/TreeView.cs#L411-L423

It force set all created TreeViewItem IsSelected to false and update Item property IsSelected to false.

and belows 420 line try reset selected elements, but there ware cannot find TreeViewItem.
> Cannot find TreeViewItem
Caused by TreeViewItem.(ItemsControl.Presenter) is null at present.
https://github.com/AvaloniaUI/Avalonia/blob/934c0f834946035d9b9bfd873739cf5c880de868/src/Avalonia.Controls/ItemsControl.cs#L266
